### ISSUE TITLE
Make sure workspace documentation matches HDF5 file

### ISSUE THREAD

**Issue by emthompson-usgs** [usgs/groundmotion-processing#381]
It looks like there's a few inconsistencies between our current implementation of the workspace structure and how we have it documented. Some minor issues like an attribute being `WaveFormMetrics` vs `WaveformMetrics`, but also we did not implement `TAG` and `WTAG` as described in the docs.

**Comment by baagaard-usgs** [comment]
I see a few issues with the current dataset names in the HDF5 files:

* The datasets in the `StationMetrics`, `TraceProcessingParameters`, and `WaveformMetrics` groups have names of the form `EVENTID_STA_TAG`. I think we need to change this to `EVENTID_NET.STA_TAG` to insure that we don't end up with duplicate names for stations with the same station code but different network codes.

* The `Provenance` group has dataset names have the station code listed twice. Currently, they are of the form `EVENTID_STA_TAG_NET_STA_CHAN`. It seems like `EVENTID_NET.STA.CHAN_TAG` or `NET.STA.CHAN_EVENTID_TAG` would be cleaner.

* The `Waveforms` group also has dataset names with the station code listed twice. Currently, they are of the form `NET.STA.LOC.CHAN__START__END__EVENTID_STA_TAG`. I think we want `NET.STA.LOC.CHAN__START__END__EVENTID_TAG`.

The bottom line is that it looks like we should be careful about our labels and tags so that we ensure we don't end up with duplicate names or the same information appearing multiple times in the dataset names.

@mhearne-usgs Are my recommendations doable without changing the ASDF API?

**Comment by emthompson-usgs** [comment]
Thanks @baagaard-usgs, @mhearne-usgs and I are trying to revisit/review this stuff now. What I am realizing is that we should add support for a "gmprocess ASDF format version." The idea that we will have a single structure and that it will never change is probably unrealistic.

**Comment by baagaard-usgs** [comment]
## Glossary
label — assigned by gmprocess user, uniquely identifies processing parameters. 
tag — assigned automatically by gmprocess to each waveform. 
path — HDF grouping. In PyASDF notation:
For Auxiliary data, path is: AuxiliaryData/DataType/”path”
In PyASDF users does not have control over the waveform path, only can set “tag”.
NET — network. 
CHA — channel. 
LOC — location code. 
EVENTID — event id. 
INST — first two characters of channel code, e.g., HN, or BH.

## Dataset naming scheme
```
Waveforms/NET.STA.LOC.CHAN__START__END__EVENTID_LABEL
AuxiliaryData/StationMetrics/EVENTID_NET.STA.LOC.INST
AuxiliaryData/WaveformMetrics/EVENTID_NET.STA.LOC.INST_LABEL
Provenance/EVENTID_NET.STA.LOC.CHAN_LABEL
```

**Comment by baagaard-usgs** [comment]
@mhearne-usgs Currently the code has the path for the waveform metrics as `/AuxiliaryData/WaveFormMetrics`. Usually we consider waveform a single word, so I think we should change it to `/AuxilairyData/WaveformMetrics`, which is consistent with the documentation.

**Comment by baagaard-usgs** [comment]
@mhearne-usgs @emthompson-usgs I hesitate to throw another wrench into the mix, but the ordering of variables within the names is not consistent. The waveform dataset names are of the form `NET.STA.LOC.CHA ... EVENTID_LABEL` but the AuxiliaryData dataset names are all of the form `EVENTID_NET.STA ... LABEL`. Since we are changing the tags in the code anyway, can we update them to have a consistent ordering.

I propose the following:
```
Waveforms/NET.STA.LOC.CHAN__START__END__EVENTID_LABEL
AuxiliaryData/StationMetrics/NET.STA.LOC.INST_EVENTID
AuxiliaryData/WaveformMetrics/NET.STA.LOC.INST_EVENTID_LABEL
Provenance/NET.STA.LOC.CHAN_EVENTID_LABEL
```

**Comment by mhearne-usgs** [comment]
@baagaard-usgs @emthompson-usgs I can live with that - I haven't actually started working on these changes yet, as there's another project I'm trying to get done first.

**Comment by emthompson-usgs** [comment]
I'm good with this also.

**Comment by mhearne-usgs** [comment]
We also have StreamProcessingParameters, which I propose to be:
AuxiliaryData/StreamProcessingParameters/NET.STA.LOC.INST_EVENTID_LABEL

**Comment by mhearne-usgs** [comment]
...and TraceProcessingParameters, which I propose to be:
AuxiliaryData/TraceProcessingParameters/NET.STA.LOC.CHA_EVENTID_LABEL

**Comment by mhearne-usgs** [comment]
@emthompson-usgs  @baagaard-usgs 

A wrinkle I forgot about: ASDF is kind of picky about what characters can go into the names/paths of provenance and auxiliary data. Specifically, those need to validate against this regex:
'^[0-9a-z][0-9a-z_]*[0-9a-z]$'

This means:
  - No uppercase
  - No hyphens
  - No periods
  - Underscores are OK

I propose:
 - replacing periods with underscores
 - replacing the double dashes '--' we've been using for blank locations with an empty string
 - lowercasing everything.

For example, this would turn this provenance name:
'BO.CHB003.--.HN2_usb000syza_processed'
 into
'bo_chb003__hn2_usb000syza_processed'

**Comment by baagaard-usgs** [comment]
I think we should contact the ASDF developers about this one. Not allowing uppercase and periods seems overly restrictive given that uppercase letters are in the group names and periods are in the dataset names for waveforms.

**Comment by emthompson-usgs** [comment]
@baagaard-usgs I think we have a temporary solution with @mhearne-usgs's PR #406, which also fixes the inefficient storage of arrays that you pointed out. Can you update the workspace docs to reflect the naming system that accommodates the ASDF restrictions?

**Comment by baagaard-usgs** [comment]
@mhearne-usgs What do I need to add to the following to account for the refactoring of the trace processing parameters?

```
Waveforms/NET.STA.LOC.CHAN__START__END__EVENTID_LABEL
AuxiliaryData/StationMetrics/NET.STA.LOC.INST_EVENTID
AuxiliaryData/WaveformMetrics/NET.STA.LOC.INST_EVENTID_LABEL
AuxiliaryData/StreamProcessingParameters/NET.STA.LOC.INST_EVENTID_LABEL
AuxiliaryData/TraceProcessingParameters/NET.STA.LOC.CHA_EVENTID_LABEL
Provenance/NET.STA.LOC.CHAN_EVENTID_LABEL
```

**Comment by emthompson-usgs** [comment]
@mhearne-usgs, please double check this. I think it should be:
```
Waveforms/NET.STA.LOC.CHAN__START__END__eventid_label
AuxiliaryData/StationMetrics/net_sta_loc_inst_eventid
AuxiliaryData/WaveformMetrics/net_sta_loc_inst_eventid_label
AuxiliaryData/StreamProcessingParameters/net_sta_loc_inst_eventid_label
AuxiliaryData/TraceProcessingParameters/net_sta_loc_chan_eventid_label
Provenance/net_sta_loc_chan_eventid_label
```
Basically, strings are lowercased where they must be because of the ASDF requirement, and dots are replaced with underscores for the same reason. Also, I think we can't change the `Waveforms/` paths that have uppercase and dots to be consistent with the rest because that is set automatically by pyasdf.

**Comment by baagaard-usgs** [comment]
@mhearne-usgs @emthompson-usgs In my ASDF layout diagram, I createed NET.STA groups under `WaveformMetrics` and `StationMetrics` for consistency with `Waveforms`. I only now noticed that this is not what was implemented. I strongly prefer consistency in the ASDF layout when at all possible. It looks like pyasdf does support this. Is there any obvious downside (other than another change to the code) in grouping the metrics by station?

**Comment by mhearne-usgs** [comment]
@baagaard-usgs Can you point to the pyasdf documentation that I could use to create a NET.STA group for auxiliary arrays?

**Comment by baagaard-usgs** [comment]
Here is the [ASDF layout diagram](https://github.com/usgs/groundmotion-processing/files/3535471/asdf_layout.pdf). I am working on adding  `TraceProccessingParameters` and `StreamProcessingParameters` to the diagram right now. I think those groups need `NET.STA` groups too (basically all the groups under `AuxiliaryData`).

What is the naming scheme for the stuff you pulled out of `TraceProcessingParameters` and into separate datasets?

**Comment by mhearne-usgs** [comment]
I think I wasn't clear - when I was asking for documentation, I wanted a pointer to documentation from the pyasdf project that would show me how to put an auxiliary array under a group. The only (scant) documentation I know of for adding these arrays is here: http://seismicdata.github.io/pyasdf/tutorial.html#adding-auxiliary-data

The paths for the "AuxArrays" are the same as TraceProcessingParameters.

**Comment by baagaard-usgs** [comment]
Did you see the API documentation at
https://seismicdata.github.io/pyasdf/asdf_data_set.html?

**Comment by mhearne-usgs** [comment]
@baagaard-usgs Nope, I didn't. :) I'll work on using the slashes to separate the paths hopefully later today.

**Comment by baagaard-usgs** [comment]
Here is the [extended ASDF layout](https://github.com/usgs/groundmotion-processing/files/3535947/asdf_layout.pdf) updated for the restrictions on tags (lowercase without periods) and the addition of `TraceProcessingParameers`, `StreamProcessingParameters`, and `AuxArrays`.

**Comment by baagaard-usgs** [comment]
@mhearne-usgs The pyasdf code seems to accept lowercase and uppercase letters and underscores (https://github.com/SeismicData/pyasdf/blob/master/pyasdf/asdf_data_set.py#L597). Am I missing something?

**Comment by baagaard-usgs** [comment]
I submitted an issue to pyasdf inquiring about regular expression validation of tag paths (https://github.com/SeismicData/pyasdf/issues/56).

**Comment by baagaard-usgs** [comment]
Created test to verify layout and naming scheme in PR https://github.com/usgs/groundmotion-processing/pull/418.
